etna/lib/etna/clients/metis/workflows/metis_upload_workflow.rb

```diff
@@ -0,0 +1,137 @@
+require 'ostruct'
+require 'digest'
+require 'fileutils'
+require 'tempfile'
+require 'securerandom'
+
+module Etna
+  module Clients
+    class Metis
+      class MetisUploadWorkflow < Struct.new(:metis_client, :metis_uid, :project_name, :bucket_name, :max_attempts, keyword_init: true)
+
+        def initialize(args)
+          super({max_attempts: 3, metis_uid: SecureRandom.hex}.update(args))
+        end
+
+        def do_upload(source_file, dest_path, &block)
+          upload = Upload.new(source_file: source_file)
```

Comment:
This workflow encapsulates the entire upload workflow, including retry & creating the containing directory if necessary.

Reply:
Most of this log is adapted from metis_client.